scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala

```diff
@@ -0,0 +1,162 @@
+package org.jetbrains.plugins.scala.lang.refactoring.move.anonymousToInner
+
+import com.intellij.java.refactoring.JavaRefactoringBundle
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.command.CommandProcessor
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.NlsContexts
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.refactoring.util.VariableData
+import com.intellij.util.IncorrectOperationException
+import org.jetbrains.plugins.scala.ScalaBundle
+import org.jetbrains.plugins.scala.lang.parser.ScalaElementType
+import org.jetbrains.plugins.scala.lang.psi.api.expr.{ScNewTemplateDefinition, ScReferenceExpression}
+import org.jetbrains.plugins.scala.lang.psi.api.statements.params.ScParameter
+import org.jetbrains.plugins.scala.lang.psi.api.toplevel.templates.ScExtendsBlock
+import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.{ScClass, ScObject, ScTemplateDefinition}
+import org.jetbrains.plugins.scala.lang.psi.api.{ScFile, ScalaPsiElement, ScalaRecursiveElementVisitor}
+import org.jetbrains.plugins.scala.lang.psi.impl.ScalaPsiElementFactory.createElementFromText
+import org.jetbrains.plugins.scala.lang.refactoring.extractMethod.ScalaVariableData
+import org.jetbrains.plugins.scala.lang.refactoring.move.anonymousToInner.ScalaAnonymousToInnerHandler.getRefactoringName
+import org.jetbrains.plugins.scala.project.ProjectContext
+
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+class ScalaAnonymousToInnerHandler {
```

Comment:
My solution is basically to add the referenced vars and declare them as `var` in the parameter declaration of the class so that they could be mutated inside the newly created class.

Reply:
@leonbur 
Oh, sorry, I missed this update.
> My solution is basically to add the referenced vars and declare them as var in the parameter declaration of the class so that they could be mutated inside the newly created class.

Unfortunately, this is not a correct solution.
Such refactoring breaks the semantics of the code, which is definitely not expected from the "refactoring"

For example, suppose you have this code:
```scala
trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar = 0
    val t = new MyTrait {
      override def foo(): Unit = {
        localVar += 1
      }
    }
    t.foo()
    t.foo()
    println(localVar)
  }
}
```
it outputs `2`

After you apply the refactoring the code becomes as follows:
```scala
trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar = 0
    val t = new MyMyTrait(localVar)
    t.foo()
    t.foo()
    println(localVar)
  }

  class MyMyTrait(var localVar: Int) extends MyTrait {
    override def foo(): Unit = {
      localVar += 1
    }
  }
}
```
and the output is `0`, because the original `var` is not changed.

Reply:
When I wrote
> Handling write operations to var could be handled in theory, but that would require a more complex design and implementation.

I kept in mind another solution.
You would need to create a `Ref` wrapper for the var.
That's what actually Scala compiler automatically does in the JVM byte code.
This is the decompiled version:
<img width="607" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/8dfd8802-e7be-49c3-b46b-936af5fba7ed">

However I am not 100% sure that we should do the same.
Users might want to address this issue in some different way.
We might leave `localVar` an ordinary parameter.
The resulting code won't compile in the places where `localVar` is updated, but it's better than silently breaking the code.
We can add some special line comment in every `localVar =`, something like
`// refactor your code or use scala.runtime.ObjectRef`
(we the actual infer `*Ref` name from the variable type)

Ideally, we could add some setting in the dialog in order user can do it automatically.
But the UX needs more thinking.
For this first version it would be OK to do as described above.

Reply:
Also, this part is still unresolved:
> It captures class parameters of the class it extracts the class (classParameter, classParameterField).
>...
>Note that in Java implementation it shows an extra option to create "static" class if there are no usages of class fields.
In future, in Scala refactoring, we could add an option to create the class in the companion object.
In that case it would capture class fields/parameters as well.
But it's fine if we don't have it in the first version.

<img width="611" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/2248167d-3b5d-47d4-ba1d-8a010e42f907">

Reply:
ok, now understand. I'll add a comment and remove the `var` prefix for writable parameters.

regarding fieldVal,classParameter,... do you want them to NOT become parameters of the created class?

Reply:
> regarding fieldVal,classParameter,... do you want them to NOT become parameters of the created class?

Yes, because those will be accessible in the extracted inner class automatically.

Reply:
The new line is added in the wrong place for the comment
<img width="451" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/8dacd35e-95c9-4e04-9458-7de11a748854">

Other thoughts:
1. Let's prepend "todo"
2. Let's mention the var names in order it's clear what that comment is about 

Something like
```scala
//todo: refactor your code or use scala.runtime.ObjectRef (localVar)
//todo: refactor your code or use scala.runtime.ObjectRef (localVar1, localVar2)
```